components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt

```diff
@@ -135,6 +135,11 @@ interface HistoryStorage : Storage {
      * Prune history storage, removing stale history.
      */
     suspend fun prune()
+
+    /**
+     * Registers storage maintenance worker that prunes database when its size exceeds a size limit.
+     * */
+    fun registerStorageMaintenanceWorker()
```

Comment:
Since we're adding a "register", would it be valuable to have an "unregister" as well? I'm not sure where we would use it exactly in the app off the top of my head. Curious to hear your opinion/thoughts.

Reply:
Good question! There is no external unregister method for this usecase which can be triggered from outside to cancel the job or to do any other operation that may mean "unregister". As this is a periodic task that we want to cancel it under particular circumstances such as:
- When constraints are not met anymore
- When there is an exception while the worker is processing
- Other than those two cases, there maybe an OS related cancellation of the request which happens when the app is killed by the OS and it cancels the work requests associated with that app. Apart from those, I cannot see a case where we would cancel/unregister/stop the worker. 

Except the OS related case, we are dealing with the cancellation of the request in the `catch` block inside the worker classes `doWork` method. For all three cases, the work request will be cancelled and a new worker will be enqueued in the next application's `onCreate` call.

The only unregister usecase I can imagine is when we want to get rid of this worker, i.e. releasing a new app rollout that will cancel all the storage maintenance work requests from the user's devices. If you think that we can make it available from the API now, I can include an unregister method that will not be used at the moment as in the issue scope I do not see anything related to cancelling the request atm.

Reply:
To be more clear and to picture the use case:
AddonUpdater has a work request register/unregister mechanism. 
When an AddonManager install an addon, the register method is called that creates the work request. And when an Addon is removed/uninstralld, that work request is unregistered/cancelled. 
With the current storage maintenance worker, there was no case mentioned that could correspond to the same behaviour of removal/uninstalling/unregistering of the work request.

Reply:
I agree - I was on a similar thought process as well. Places is not explicitly related to Sync, but if it was, we could at least convince ourselves that we should perform some register/unregister actions based on sync sign in/out. Having the periodic task always enqueued doesn't seem to have an negative repercussions as far as I can see.

We have telemetry (via app services) to know how to react to this feature landing, so we can re-evaluate based on that information.

Thanks for humouring me here. 🙂